### ISSUE TITLE
[Proposal] Separation of a "core" set of classes from the "standard library"

### ISSUE THREAD

**Issue by js-ojus** [crystal-lang/crystal#1795]
Arguably, we can have a "core" set of classes.  This core is defined as part of the language, understood by the compiler, and is hence available in macros.  Others can be moved into the standard library.

Core classes are available at compile-time and at run-time.  Classes in the standard library are available at run-time, but not at compile-time.

Core classes are self-contained; at worst, they can depend on other core classes.  Standard library classes can depend on core classes; they can, of course, depend on other standard library classes too.

Does this make sense?

**Comment by ysbaddaden** [comment]
Yes, it does, and it already exists. The corelib is what is required by `src/prelude.cr` and the stdlib is everything else. There already is a `spec/std` folder too, but no matching `src/std`.

Having `src/compiler`, `src/core` and `src/std` could be nice.

**Comment by ozra** [comment]
+1

**Comment by asterite** [comment]
This same question was asked in Google Groups and Reddit. This is my answer:

https://www.reddit.com/r/crystal_programming/comments/3p1juq/crystal_090_released/cw7sgoo

> A quick search for "elixir json" yields three projects:
https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=elixir%20json
This doesn't look like a problem at first: in my project I use the JSON library of my convenience. The problem is when you build a library and you have to choose one, and then your project requires multiple libraries that depend on different implementation.
For things so standard and ubiquitous as HTTP, XML, YAML and JSON, I think the best is to have good and efficient implementations in the standard library, to improve compatibility between different libraries.

Also @waj's [reply to a different issue](https://github.com/manastech/crystal/issues/1710#issuecomment-149885252) applies here too.

In short: yes, we can have different repositories for these libraries, but we'd like the standard library to come with them and to avoid having many incompatible implementations of the same feature.

**Comment by ozra** [comment]
The standard library should definitely come along with the compiler - afaic it's just a clarification of putting core-libs in specific dirs, not intermixing with the others, _file-wise_. Or did I miss something in this issue? That was my intention anyway.

**Comment by ysbaddaden** [comment]
I understand this the same as you @ozra , this is more about distinguishing more easily what's part of the corelib (ie. automatically required) and the stdlib (manually required) upfront. Either by looking at files in the source code, or in the documentation.

**Comment by asterite** [comment]
What's a specific proposal to do this? Something that will make it clear, either in the docs or the API, that some classes are automatically included in every program.

**Comment by straight-shoota** [comment]
Closing stale proposal. There has been no comment activity for nearly five years.

I don't think this is a problem. Core features just work out of the box. Non-core features need an explicit `require` and the documentation makes sure to show that at least in the examples. Maybe this could be showed more explicitly in the API docs. But lack of interest in this issue suggests it does not pose a problem in practice.
